exasol/shared_memory_vault.py

```diff
@@ -0,0 +1,248 @@
+           creation_time: datetime) -> Tuple[int, bytearray]:
+    """
+    Creates a bytearray with encoded content and its creation datetime.
+    Currently, the content is not being encrypted. It gets appended by the
+    Cyclic Redundancy Check (CRC) code. The CRC is computed over both the timestamp
+    and the content.
+
+    The need for a CRC is debatable. Its use is motivated by the problem of
+    non-synchronised concurrent access to the shared memory. In theory, the content
+    can get corrupted by simultaneous reading/writing. Hence, is the need to validate
+    the data retrieved from the shared memory. On the other hand, given the use case,
+    the possibility of concurrent access is only hypothetical. The implication of
+    reading an impaired data is insignificant and CRC cannot guarantee the validity
+    with 100% accuracy anyway.
+    """
+
+    # Prepend the content by the timestamp and convert the whole thing to a binary
+    # sequence.
+    ts_bytes = struct.pack('d', creation_time.timestamp())
+    body = ts_bytes + content.encode('utf8')
+    bin_body = _bytes_to_bin(body)
+
+    # Compute the CRC of the content right-padded with n zeros.
+    divisor_size = _get_divisor_size(crc_divisor)
+    padding = '0' * len(crc_divisor)
+    bin_cr = compute_crc(bin_body + padding, crc_divisor)
+    cr = int(bin_cr, 2)
+
+    # Put together the content bytes and the CRC bytes.
+    cont_size = len(body) + divisor_size
+    enc_content = bytearray(cont_size)
+    enc_content[:-divisor_size] = body
+    enc_content[-divisor_size:] = cr.to_bytes(divisor_size, byteorder='big',
+                                              signed=False)
+    return cont_size, enc_content
+
+
+def decode(enc_content: bytearray, crc_divisor: str) -> Tuple[bool, datetime, str]:
+    """
+    Decodes and validates a content encoded in a bytearray.
+    Returns the validity flag, the datetime of the content creation and the textual
+    content. If the CRC code is invalid the function returns (False, <any datetime>,
+    ''). Otherwise, it returns (True, <content creation datetime>, <textual content>).
+    """
+
+    # Compute the CRC code of the content that should include its own CRC code.
+    divisor_size = _get_divisor_size(crc_divisor)
+    bin_body = _bytes_to_bin(enc_content[:-divisor_size])
+    bin_cr = _bytes_to_bin(enc_content[-divisor_size:])[-len(crc_divisor):]
+    bin_cr = compute_crc(bin_body + bin_cr, crc_divisor)
+
+    # For a valid content the computed CRC should be zero.
+    if int(bin_cr, 2) == 0:
+        divisor_size = _get_divisor_size(crc_divisor)
+        # Decode the content creation timestamp.
+        ts = struct.unpack('d', enc_content[:8])[0]
+        # Decode the content.
+        content = enc_content[8:-divisor_size].decode('utf8')
+        return True, datetime.fromtimestamp(ts), content
+    return False, datetime.min, ''
+
+
+def _open_shared_memory(storage_name: str, max_size: int,
```

Comment:
I now also remembered the issue with shared memory and docker. All is nice and good, if you assume the default behavior of docker regarding ipc namespace. However, there exist the option --ipc=host, in that case the shared memory of the host is used and not cleared during docker stop. We need at least a warning in the docs and the entry point process needs try to remove it when stopping. With a process this would be given and we only would need to care about clearing

Reply:
And, I had an idea how we don't need to care about this. For reading we can create a session copy of the DB with a session password.